### PR TITLE
Adds to and toMap to SourceWithContext

### PR DESCRIPTION
**Pull request by patriknw**
Rebase of https://github.com/akka/akka/pull/26412
also adding `runWith` to `SourceWithContext`

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceWithContextSpec.scala [hunk 1]
```diff
@@ -21,8 +21,8 @@ class SourceWithContextSpec extends StreamSpec {
       val msg = Message("a", 1L)
       Source(Vector(msg))
         .asSourceWithContext(_.offset)
-        .asSource
-        .runWith(TestSink.probe[(Message, Long)])
+        .toMat(TestSink.probe[(Message, Long)])(Keep.right)
+        .run
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceWithContextSpec.scala]: Is there still test coverage of `asSource` elsewhere?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceWithContextSpec.scala [hunk 2]
```diff
@@ -93,8 +91,8 @@ class SourceWithContextSpec extends StreamSpec {
           List(1, 2, 3, 4).map(i ⇒ s"$str-$i")
         }
         .grouped(2)
-        .asSource
-        .runWith(TestSink.probe[(Seq[String], Seq[Long])])
+        .toMat(TestSink.probe[(Seq[String], Seq[Long])])(Keep.right)
+        .run
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceWithContextSpec.scala]: Should we leave one test actually using `.asSource` directly?

**Review by johanandren — APPROVED** [review]
LGTM with one test consideration.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceWithContextSpec.scala (hunk 1 above)
Comment (patriknw) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceWithContextSpec.scala]: There is "be able to get turned back into a normal Source"

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceWithContextSpec.scala (hunk 2 above)
Comment (patriknw) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/SourceWithContextSpec.scala]: ^

**Pull request merged**